src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,328 @@ public static CFProxySettings GetSystemProxySettings ()
+				} finally {
+					// clean resources
+					Marshal.FreeHGlobal (pacDataPtr); 
+				}
+			}, cancellationToken).ConfigureAwait (false);
+
+			return (proxies: proxies, error: outError);
+		}
+
+		static CFProxy[] ExecutePacCFRunLoopSourceBlocking (CreatePACCFRunLoopSource factory, out NSError outError)
+		{
+			var runLoop = CFRunLoop.Current;
+			CFProxy [] proxies = null;
+			outError = null;
+
+			// build a struct that will have all the needed info for the callback
+			var pacCbData = new PACProxyCallbackData ();
+			pacCbData.CFRunLoopPtr = runLoop.Handle;
+			var pacDataPtr = Marshal.AllocHGlobal (Marshal.SizeOf(pacCbData));
+			try {
+				Marshal.StructureToPtr (pacCbData, pacDataPtr, false); 
+
+				var clientContext = new CFStreamClientContext ();
+				clientContext.Info = pacDataPtr;
+
+				using (var loopSource = new CFRunLoopSource (factory (ExecutePacCallback, ref clientContext)))
+				using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+					runLoop.AddSource (loopSource, mode);
+					runLoop.RunInMode (mode, double.MaxValue, false);
+					runLoop.RemoveSource (loopSource, mode);
+				}
+				pacCbData = (PACProxyCallbackData) Marshal.PtrToStructure(pacDataPtr, typeof (PACProxyCallbackData));
+				// get data from the struct
+				if (pacCbData.Error != null)
+					outError = pacCbData.Error;
+				else 
+					proxies = pacCbData.ProxyList;
+				// clean resources
+				return proxies;
+			} finally {
+				Marshal.FreeHGlobal (pacDataPtr); 
+			}
+		}
+
+		// helper struct to contain all the data that is needed to execute the user cb
+		struct PACProxyUserCallbackData
+		{
+			public CFProxyAutoConfigurationResultCallback UserCallback; // Pointer to the user callback
+			public IntPtr UserClientDataPtr; // Pointer to the passed objc from the user
+		}
+
+		[MonoPInvokeCallback (typeof (CFProxyAutoConfigurationResultCallbackInternal))]
+		static void ExecutePacUserCallback (IntPtr client, IntPtr proxyListPtr, IntPtr errorPtr)
+		{
+			// grab the required structure and set the data, according apple docs:
+			// client
+			// The client reference originally passed in the clientContext parameter of the 
+			// CFNetworkExecuteProxyAutoConfigurationScript or CFNetworkExecuteProxyAutoConfigurationURL call
+			// that triggered this callback.
+			// Well, that is NOT TRUE, the client passed is the client.Info pointer not the client.
+			var pacCbData =  (PACProxyUserCallbackData) Marshal.PtrToStructure (client, typeof (PACProxyUserCallbackData));
+			// convert the data to be used for the user
+			CFProxy [] proxies = null;
+			NSError error = null;
+			if (errorPtr != IntPtr.Zero) {
+				error = Runtime.GetNSObject<NSError> (errorPtr);
+			} else {
+				// retain so that is nore released
+				CFObject.CFRetain (proxyListPtr);
+				proxies = ParseProxies (proxyListPtr);
+			}
+			// get the callback to be executed
+			pacCbData.UserCallback (Runtime.GetNSObject<NSObject> (pacCbData.UserClientDataPtr), proxies, error);
```

Comment:
I think the above can be simplified to:

```csharp
var proxies = ParseProxies (proxyListPtr);
var error = (NSError) Runtime.GetNSObject (errorPtr);
pacDbData.UserCallback (Runtime.GetNSObject (pacCbData.UserClientDataPtr), proxies, error);
```

Reply:
I did it this way following @spouliot comment here: https://github.com/xamarin/xamarin-macios/pull/6201#discussion_r303969327